### PR TITLE
Blonde monkey 11051

### PR DESCRIPTION
**Pull request by fp-ps**

### CODE REVIEW

File: backend/de.metas.util.web/src/main/java/de/metas/util/web/audit/ApiAuditService.java [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+	{
+		final OrgId orgId = apiAuditConfig.getOrgId();
+
+		final CustomHttpRequestWrapper httpRequest = new CustomHttpRequestWrapper(request, objectMapper);
+
+		final ApiRequestAudit requestAudit = logRequest(httpRequest, apiAuditConfig.getApiAuditConfigId(), orgId);
+
+		final ApiAuditLoggable apiAuditLoggable = createLogger(requestAudit.getIdNotNull());
+
+		try (final IAutoCloseable loggableRestorer = Loggables.temporarySetLoggable(apiAuditLoggable))
```
Comment (metas-ts) [inline backend/de.metas.util.web/src/main/java/de/metas/util/web/audit/ApiAuditService.java:139]: i didn't yet dig into CompletableFuture, but are we sure that `apiAuditLoggable` will be in the threat in which the request will be processed? on my first glimpse i think the executor in HttpCallScheduler won't have the loggable..

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/ApiAuditLoggable.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+			flush();
+		}
+
+		return this;
+	}
+
+	@Override
+	public void flush()
+	{
+		final List<ApiRequestAuditLog> logEntries = buffer;
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/ApiAuditLoggable.java:98]: I assume `addLog` and `flush` are not called from different threads. right?

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/ApiAuditLoggable.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+			apiAuditRequestLogDAO.insertLogs(logEntries);
+		}
+		catch (final Exception ex)
+		{
+			// make sure flush never fails
+			logger.warn("Failed saving {} log entries but IGNORED: {}", logEntries.size(), logEntries, ex);
+		}
+	}
+
+	private ApiRequestAuditLog createLogEntry(@NonNull final String msg, final Object... msgParameters)
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/ApiAuditLoggable.java:117]: static

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/HttpMethod.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+			return null;
+		}
+
+		return ofCode(code);
+	}
+
+	@NonNull
+	public static Optional<HttpMethod> ofCodeOptional(@Nullable final String code)
+	{
+		if (Check.isBlank(code))
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/HttpMethod.java:75]: pls use ReferenceListAwareEnums.index(...).
e.g.
```java
	private static final ReferenceListAwareEnums.ValuesIndex<TooltipType> index = ReferenceListAwareEnums.index(values());
```

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/config/ApiAuditConfigRepository.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+@Repository
+public class ApiAuditConfigRepository
+{
+	private final IQueryBL queryBL = Services.get(IQueryBL.class);
+
+	public ImmutableList<ApiAuditConfig> getAllConfigsByOrgId(@NonNull final OrgId orgId)
+	{
+		return queryBL.createQueryBuilder(I_API_Audit_Config.class)
+				.addOnlyActiveRecordsFilter()
+				.addEqualsFilter(I_API_Audit_Config.COLUMNNAME_AD_Org_ID, orgId)
```
[the remark this answers is not in the document]
Comment (teosarca) [reply backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/config/ApiAuditConfigRepository.java:45]: more, I would consider introducing `ApiAuditConfigMap` which stores ALL configs, so I will retrieve them once.

Regarding returning `ImmutableList<ApiAuditConfig>`, consider introducing and returning `ApiAuditConfigList`, which wraps that list and provides some methods for picking the right Config based on HTTP method and operation name.

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/config/ApiAuditConfigRepository.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+				.create()
+				.list()
+				.stream()
+				.map(this::buildConfigFromRecord)
+				.collect(ImmutableList.toImmutableList());
+	}
+
+	@NonNull
+	private ApiAuditConfig buildConfigFromRecord(@NonNull final I_API_Audit_Config record)
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/config/ApiAuditConfigRepository.java:54]: static?

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/config/NotificationTriggerType.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import lombok.Getter;
+import lombok.NonNull;
+import org.adempiere.exceptions.AdempiereException;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+
+import static org.compiere.model.X_API_Audit_Config.NOTIFYUSERINCHARGE_AllenAufrufen;
+import static org.compiere.model.X_API_Audit_Config.NOTIFYUSERINCHARGE_AufrufenMitFehler;
+import static org.compiere.model.X_API_Audit_Config.NOTIFYUSERINCHARGE_Niemals;
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/config/NotificationTriggerType.java:37]: pls trl those to English, i.e. set AD_Ref_List.ValueName.

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/request/Status.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import lombok.Getter;
+import lombok.NonNull;
+import org.adempiere.exceptions.AdempiereException;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+
+import static org.compiere.model.X_API_Request_Audit.STATUS_Empfangen;
+import static org.compiere.model.X_API_Request_Audit.STATUS_Fehler;
+import static org.compiere.model.X_API_Request_Audit.STATUS_Verarbeitet;
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/request/Status.java:37]: pls trl to EN

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/response/ApiResponseAuditRepository.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	@NonNull
+	public ApiResponseAudit getById(@NonNull final ApiResponseAuditId apiResponseAuditId)
+	{
+		final I_API_Response_Audit record = InterfaceWrapperHelper.load(apiResponseAuditId, I_API_Response_Audit.class);
+
+		return recordToResponseAudit(record);
+	}
+
+	@NonNull
+	private ApiResponseAudit recordToResponseAudit(@NonNull final I_API_Response_Audit record)
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/audit/response/ApiResponseAuditRepository.java:63]: static

File: backend/de.metas.util.web/src/main/java/de/metas/util/web/audit/ApiAuditService.java [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+			@NonNull final ApiAuditConfig apiAuditConfig,
+			@NonNull final ApiRequestAudit apiRequestAudit,
+			@NonNull final CompletableFuture<ApiResponse> actualRestApiResponseCF,
+			@NonNull final HttpServletResponse httpServletResponse) throws IOException, InterruptedException, ExecutionException, TimeoutException
+	{
+		final Object actualRestApiResponseBody;
+		final int httpStatus;
+		if (apiAuditConfig.isInvokerWaitsForResponse())
+		{
+			final ApiResponse actualRestApiResponse = actualRestApiResponseCF.get(120, TimeUnit.SECONDS); // todo establish a reasonable value
```
Comment (teosarca) [inline backend/de.metas.util.web/src/main/java/de/metas/util/web/audit/ApiAuditService.java:325]: might be good to have it configurable via ApiAuditConfig?

File: backend/de.metas.util.web/src/main/java/de/metas/util/web/filter/ApiAuditFilter.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+			{
+				chain.doFilter(request, response);
+				return;
+			}
+
+			apiAuditService.processHttpCall(httpServletRequest, httpServletResponse, matchingAuditConfig.get());
+		}
+		catch (final Throwable t)
+		{
+			httpServletResponse.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, t.getLocalizedMessage());
```
Comment (teosarca) [inline backend/de.metas.util.web/src/main/java/de/metas/util/web/filter/ApiAuditFilter.java:74]: don't we have to log the stacktrace or somehow to send it forward?

File: backend/de.metas.util.web/src/main/java/de/metas/util/web/filter/ApiAuditFilter.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Optional;
+
+public class ApiAuditFilter implements Filter
```
Comment (teosarca) [inline backend/de.metas.util.web/src/main/java/de/metas/util/web/filter/ApiAuditFilter.java:39]: i guess package level is enough

**Review by metas-ts — APPROVED** [review]
lgtm. i'm aware that you didn't get to act on most comments from me&teo.
please plan to react to them in the coming api-audit-issue

**Pull request merged**